local/network.go

```diff
@@ -533,18 +547,22 @@ func (ln *localNetwork) GetAllNodes() (map[string]node.Node, error) {
 }
 
 func (ln *localNetwork) Stop(ctx context.Context) error {
-	ln.lock.Lock()
-	defer ln.lock.Unlock()
+	err := network.ErrStopped
+	ln.stopOnce.Do(
```

Comment:
Just a comment on the pattern. I shall prefer something similar to this, but with the error to return after the starting execution included in the `once` call. Somethings like this.
```
return ln.stopOnce(func, afterOnceErr)
```
Of course I understand this does not exits but shall more readable in my opinion.

Reply:
I'm not sure I understand what you mean. Maybe you could paste a snippet with the code you're thinking of?

If you're suggesting that we return from `Stop` inside the anonymous function in `Do`, that isn't possible because returning inside that anonymous function will only return from the anonymous function, not from `Stop`

Reply:
Ohh something like these 

```
type doOnce struct {                                                                                                        
    mu sync.Mutex                                                                                                           
    done bool                                                                                                               
}                                                                                                                           
                                                                                                                            
func (do doOnce) do(toDo func() error, errAfterOnce error) error {                                                          
    do.mu.Lock()                                                                                                            
    defer do.mu.Unlock()                                                                                                    
    var err error                                                                                                           
    if !do.done {                                                                                                           
        err = toDo()                                                                                                        
        do.done = true                                                                                                      
    } else {                                                                                                                
        err = errAfterOnce                                                                                                  
    }                                                                                                                       
    return err                                                                                                              
}                           
```
Then for field 

`stopOnce doOnce`

Using

 ```
func (ln *localNetwork) Stop(ctx context.Context) error {                                                                   
    return ln.stopOnce.Do(                                                                                                  
        func() error {                                                                                                      
            close(ln.onStopCh)                                                                                              
            ln.lock.Lock()                                                                                                  
            defer ln.lock.Unlock()                                                                                          
            return ln.stop(ctx)                                                                                             
        },                                                                                                                  
        network.ErrStopped,                                                                                                 
    )                                                                                                                       
}              
```

Reply:
I understand this really only works for error returns, so probably not general

Reply:
I think it's cleaner as it is